tool/internal/instrument/optimize.go

```diff
@@ -110,55 +110,59 @@ func removeAfterTrampolineCall(tjump *TJump) error {
 	return nil
 }
 
-func populateHookContextLiteral(tjump *TJump, expr dst.Expr) {
+// newHookContextImpl constructs a new HookContextImpl structure literal and
+// populates its Params && ReturnValues field with addresses of all arguments.
+// The HookContextImpl structure is used to pass arguments to the exit trampoline
+func newHookContextImpl(tjump *TJump) dst.Expr {
 	targetFunc := tjump.target
-	elems, _ := expr.(*dst.UnaryExpr).X.(*dst.CompositeLit)
-	paramElem, returnElem := elems.Elts[0], elems.Elts[1]
-	// Populate call context literal with addresses of all arguments
-	names := make([]dst.Expr, 0)
+	structName := "HookContextImpl" + util.CRC32(tjump.rule.String())
+
+	// Build params slice: []interface{}{}
+	paramNames := make([]dst.Expr, 0)
 	for _, name := range getNames(targetFunc.Type.Params) {
-		names = append(names, ast.AddressOf(name))
+		paramNames = append(paramNames, ast.AddressOf(name))
 	}
-	paramLiteral, _ := paramElem.(*dst.KeyValueExpr).Value.(*dst.CompositeLit)
-	paramLiteral.Elts = names
-	// Populate return values
+	paramsSlice := &dst.CompositeLit{
+		Type: ast.ArrayType(ast.InterfaceType()),
+		Elts: paramNames,
+	}
+
+	// Build returnVals slice: []interface{}{}
+	returnElts := make([]dst.Expr, 0)
 	if targetFunc.Type.Results != nil {
-		rets := make([]dst.Expr, 0)
 		for _, name := range getNames(targetFunc.Type.Results) {
-			rets = append(rets, ast.AddressOf(name))
+			returnElts = append(returnElts, ast.AddressOf(name))
 		}
-		returnLiteral, _ := returnElem.(*dst.KeyValueExpr).Value.(*dst.CompositeLit)
-		returnLiteral.Elts = rets
 	}
-}
+	returnValsSlice := &dst.CompositeLit{
+		Type: ast.ArrayType(ast.InterfaceType()),
+		Elts: returnElts,
+	}
 
-// newHookContextImpl constructs a new HookContextImpl structure literal and
-// populates its Params && ReturnValues field with addresses of all arguments.
-// The HookContextImpl structure is used to pass arguments to the exit trampoline
-func newHookContextImpl(tjump *TJump) (dst.Expr, error) {
-	// TODO: This generated structure construction can also be marked via line
-	// directive
-	// One line please, otherwise debugging line number will be a nightmare
-	tmpl := fmt.Sprintf("&HookContextImpl%s{params:[]interface{}{},returnVals:[]interface{}{}}",
-		util.CRC32(tjump.rule.String()))
-	p := ast.NewAstParser()
-	astRoot, err := p.ParseSnippet(tmpl)
-	if err != nil {
-		return nil, err
+	// Build the struct literal: &HookContextImpl{params:..., returnVals:...}
+	ctxExpr := &dst.UnaryExpr{
```

Comment:
Let's reuse the functions from primitives.go as much as possible to build the AST.